phidgets/manager.go

```diff
@@ -0,0 +1,139 @@
+package phidgets
+
+/*
+#cgo CFLAGS: -g -Wall
+#cgo LDFLAGS: -lphidget22
+#include <stdlib.h>
+#include <phidget22.h>
+typedef void (*attach_fcn)(PhidgetManagerHandle man, void *ctx, PhidgetHandle channel);
+void cattach_callback(PhidgetManagerHandle man, void *ctx, PhidgetHandle channel);
+*/
+import "C"
+import (
+	"errors"
+	"fmt"
+	"time"
+	"sync"
+	"unsafe"
+)
+
+var (
+	handles []*PhidgetHandle
+	mutex   sync.Mutex
+)
+
+//PhidgetManager is the struct that is a phidget manager handle
+type PhidgetManager struct {
+	handle C.PhidgetManagerHandle
+}
+
+//PhidgetHandle is a generic Phidget handle
+type PhidgetHandle struct {
+	handle C.PhidgetHandle
+}
+
+//export attach_handler
+func attach_handler(man C.PhidgetManagerHandle, ctx unsafe.Pointer, channel C.PhidgetHandle) {
+	mutex.Lock()
+	C.Phidget_retain(channel)
+	handles = append(handles, &PhidgetHandle{handle: channel})
+	mutex.Unlock()
+}
+
+//NewPhidgetManager Create creates a phidget manager
+func NewPhidgetManager() (*PhidgetManager, error) {
+	p := &PhidgetManager{}
+	C.PhidgetManager_create(&p.handle)
+
+	cerr := C.PhidgetManager_setOnAttachHandler(p.handle, (C.attach_fcn)(unsafe.Pointer(C.cattach_callback)), nil)
+	if cerr != C.EPHIDGET_OK {
+		C.PhidgetManager_delete(&p.handle)
+		return nil, errors.New(getErrorDescription(p, cerr))
+	}
+
+	if cerr := C.PhidgetManager_open(p.handle); cerr != C.EPHIDGET_OK {
+		C.PhidgetManager_delete(&p.handle)
+		return nil, errors.New(getErrorDescription(p, cerr))
+	}
+	// Give the handles time to attach
+	time.Sleep(time.Millisecond * 10)
+
+	return p, nil
+}
+
+//ListPhidgets returns a list of phidgets that have been discovered
+func (m *PhidgetManager) ListPhidgets() []*PhidgetHandle {
+	mutex.Lock()
+	l := append([]*PhidgetHandle{}, handles...)
+	mutex.Unlock()
+	return l
+}
+
+//Close - close the handle and delete it
+func (p *PhidgetManager) Close() error {
+	if cerr := C.PhidgetManager_close(p.handle); cerr != C.EPHIDGET_OK {
+		return errors.New(getErrorDescription(p, cerr))
+	}
+	if cerr := C.PhidgetManager_delete(&p.handle); cerr != C.EPHIDGET_OK {
+		return errors.New(getErrorDescription(p, cerr))
+	}
+	return nil
+}
+
+//ChannelClass retrieves the class of channel for this handle
+func (p *PhidgetHandle) ChannelClass() string {
```

Comment:
Apparently my review wasn't submitted from a few days ago. Sorry for the delay!

Reply:
I've updated the PR, but given the discussion on #5 perhaps don't bother merging this just yet. I'll leave this here in case the other refactoring doesn't pan out well.

Reply:
Sure, I'll hold off for now. Thanks!